network/validator/authorized_sender_validator.go

```diff
@@ -20,63 +20,74 @@ var (
-
-		switch {
-		case err == nil:
-			return msgType, nil
-		case message.IsUnknownMsgTypeErr(err):
-			violation := &slashing.Violation{Identity: identity, PeerID: from.String(), MsgType: msgType, Channel: channel, IsUnicast: isUnicast, Err: err}
-			slashingViolationsConsumer.OnUnknownMsgTypeError(violation)
-			return msgType, err
-		case errors.Is(err, message.ErrUnauthorizedMessageOnChannel) || errors.Is(err, message.ErrUnauthorizedRole):
-			violation := &slashing.Violation{Identity: identity, PeerID: from.String(), MsgType: msgType, Channel: channel, IsUnicast: isUnicast, Err: err}
-			slashingViolationsConsumer.OnUnAuthorizedSenderError(violation)
-			return msgType, err
-		case errors.Is(err, ErrSenderEjected):
-			violation := &slashing.Violation{Identity: identity, PeerID: from.String(), MsgType: msgType, Channel: channel, IsUnicast: isUnicast, Err: err}
-			slashingViolationsConsumer.OnSenderEjectedError(violation)
-			return msgType, ErrSenderEjected
-		default:
-			// this condition should never happen and indicates there's a bug
-			// don't crash as a result of external inputs since that creates a DoS vector
-			log.Error().
-				Err(err).
-				Str("peer_id", from.String()).
-				Str("role", identity.Role.String()).
-				Str("peer_node_id", identity.NodeID.String()).
-				Str("message_type", msgType).
-				Bool("unicast_message", isUnicast).
-				Msg("unexpected error during message validation")
-			return msgType, err
-		}
+	msgType, err := av.isAuthorizedSender(identity, channel, msg)
+
+	switch {
+	case err == nil:
+		return msgType, nil
+	case message.IsUnknownMsgTypeErr(err):
+		violation := &slashing.Violation{Identity: identity, PeerID: from.String(), MsgType: msgType, Channel: channel, IsUnicast: isUnicast, Err: err}
+		av.slashingViolationsConsumer.OnUnknownMsgTypeError(violation)
+		return msgType, err
+	case errors.Is(err, message.ErrUnauthorizedMessageOnChannel) || errors.Is(err, message.ErrUnauthorizedRole):
+		violation := &slashing.Violation{Identity: identity, PeerID: from.String(), MsgType: msgType, Channel: channel, IsUnicast: isUnicast, Err: err}
+		av.slashingViolationsConsumer.OnUnAuthorizedSenderError(violation)
+		return msgType, err
+	case errors.Is(err, ErrSenderEjected):
+		violation := &slashing.Violation{Identity: identity, PeerID: from.String(), MsgType: msgType, Channel: channel, IsUnicast: isUnicast, Err: err}
+		av.slashingViolationsConsumer.OnSenderEjectedError(violation)
+		return msgType, ErrSenderEjected
+	default:
+		// this condition should never happen and indicates there's a bug
+		// don't crash as a result of external inputs since that creates a DoS vector
+		av.log.Error().
+			Err(err).
+			Str("peer_id", from.String()).
+			Str("role", identity.Role.String()).
+			Str("peer_node_id", identity.NodeID.String()).
+			Str("message_type", msgType).
+			Bool("unicast_message", isUnicast).
+			Msg("unexpected error during message validation")
+		return msgType, err
```

Comment:
Please also define the metric for `unexpected` error. We may later need to set an alert for this on Grafana.

Reply:
https://github.com/onflow/flow-go/pull/3146/commits/70be5b17cdd4fc8ebf79b6199459c5af46da6165